Set window scene identifier for iOS WebAuthn requests

g-davidson opened the pull request:
#### ee7418da8cb914101b157c5cf471a803ba2b1798
<pre>
Set window scene identifier for iOS WebAuthn requests
rdar://106483264

Reviewed by Aditya Keerthi.

Start setting the window scene identifier for WebAuthn requests.

* Source/WebKit/Platform/spi/Cocoa/AuthenticationServicesCoreSPI.h:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:

Canonical link: <a href="https://commits.webkit.org/263182@main">https://commits.webkit.org/263182@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9431cb889962e23301b7693c534fff24f4758bf6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3464 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3519 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3651 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4890 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3759 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3599 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3553 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/3015 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3506 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3765 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3120 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4710 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1273 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3099 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/3011 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3066 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3158 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/4472 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3530 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/2838 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3073 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3097 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/941 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3096 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3351 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/Platform/spi/Cocoa/AuthenticationServicesCoreSPI.h:
```diff
@@ -275,6 +275,10 @@ typedef NS_ENUM(NSInteger, ASCredentialRequestStyle) {
 
 @property (nonatomic, nullable, copy) ASGlobalFrameIdentifier *globalFrameID;
+
+#if TARGET_OS_IOS
```
We should add a named macro in `PlatformHave.h` for this property.

Note that the equivalent of `TARGET_OS_IOS` is `PLATFORM(IOS) || PLATFORM(MACCATALYST)`.

g-davidson replied:
This is a forward declaration of SPI and the macro matches the source declaration. Defining it differently is likely to cause us trouble down the road.

pxlcoder replied:
We generally only use `TARGET_OS_` macros in framework headers. Using `HAVE` is the approach we use in other `...SPI.h` headers.

`TARGET_OS_IOS` == `PLATFORM(IOS) || PLATFORM(MACCATALYST)` in WebKit. Many more things would break if that were to change.

@hortont424 may have thoughts.

g-davidson replied:
`HAVE()` macros would definitely be the right approach if this declaration was from a set of related declarations for a whole feature that may be unavailable on different platforms. But that's not the case here :)

This whole file is a forward declaration of a series of framework headers, and right now is textually identical to the source files it's redeclaring, which makes updates as easy as copy/paste. If we switch to using piecemeal `HAVE()` sprinkled throughout, in the best case we'd just be defining those macros to use `TARGET_OS_...` anyway (because that's how the source file declares them) and introducing a new maintenance burden to updating to this file. However introducing discrepancies in the macros used between this file and the source files makes it much easier for them to deviate and introduce bugs in the future.

pxlcoder replied:
Sounds good! It turns out we don't have strict rules about this case, so this is fine.

Though I would like to note that it may not always be as easy as copy/paste. In this case, we're fine because we don't need to support building on older iOS versions, where the SPI may not exist. However, if we're adopting newer SPI on a particular macOS version, we would need a `HAVE`.

pxlcoder commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -500,6 +501,8 @@ static inline void continueAfterRequest(RetainPtr<id <ASCCredentialProtocol>> cr
     }
 #endif // PLATFORM(MAC) || PLATFORM(MACCATALYST)
 #if PLATFORM(IOS)
```
@pascoej Is this `PLATFORM(IOS)` intentionally? Do we want to exclude Catalyst here?

g-davidson replied:
Yes this is intentional. Catalyst is covered in the condition above.

Commit:
Set window scene identifier for iOS WebAuthn requests
rdar://106483264

Reviewed by Aditya Keerthi.

Start setting the window scene identifier for WebAuthn requests.

* Source/WebKit/Platform/spi/Cocoa/AuthenticationServicesCoreSPI.h:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:

Canonical link: https://commits.webkit.org/263182@main
